File: agent-test/backend/tools.py
"""
Backend tools for AssetFlow Agent.
Each function queries the SQLite database and returns structured results.
"""
import sqlite3
import re
from datetime import datetime


def _get_conn(db_path: str) -> sqlite3.Connection:
    conn = sqlite3.connect(db_path)
    conn.row_factory = sqlite3.Row
    conn.execute("PRAGMA foreign_keys = ON")
    return conn


def _rows_to_dicts(rows) -> list:
    return [dict(r) for r in rows]


# ───────────────────────────────────────────────────────────
# 1. query_assets
# ───────────────────────────────────────────────────────────
def query_assets(db_path, category=None, plant=None, status=None,
                 department=None, is_production_critical=None):
    """Return a list of assets matching the given filters."""
    conn = _get_conn(db_path)
    sql = """
        SELECT a.id, a.asset_tag, a.name, a.category, a.status,
               a.condition, a.location, a.is_bookable, a.is_production_critical,
               p.name AS plant_name, p.city AS plant_city,
               d.name AS department_name
        FROM assets a
        JOIN plants p ON a.plant_id = p.id
        JOIN departments d ON a.department_id = d.id
        WHERE 1=1
    """
    params = []
    if category:
        sql += " AND a.category = ?"
        params.append(category)
    if plant:
        sql += " AND (p.name LIKE ? OR p.city LIKE ?)"
        params.extend([f"%{plant}%", f"%{plant}%"])
    if status:
        sql += " AND a.status = ?"
        params.append(status)
    if department:
        sql += " AND d.name LIKE ?"
        params.append(f"%{department}%")
    if is_production_critical is not None:
        sql += " AND a.is_production_critical = ?"
        params.append(int(is_production_critical))
    sql += " ORDER BY a.asset_tag LIMIT 100"

    rows = conn.execute(sql, params).fetchall()
    conn.close()
    return _rows_to_dicts(rows)


# ───────────────────────────────────────────────────────────
# 2. get_asset_detail
# ───────────────────────────────────────────────────────────
def get_asset_detail(db_path, asset_tag):
    """Full detail on a single asset including allocation and maintenance history."""
    conn = _get_conn(db_path)

    asset = conn.execute("""
        SELECT a.*, p.name AS plant_name, p.city AS plant_city,
               d.name AS department_name
        FROM assets a
        JOIN plants p ON a.plant_id = p.id
        JOIN departments d ON a.department_id = d.id
        WHERE a.asset_tag = ?
    """, (asset_tag,)).fetchone()

    if not asset:
        conn.close()
        return {"error": f"Asset '{asset_tag}' not found"}

    asset_dict = dict(asset)

    allocations = conn.execute("""
        SELECT aa.*, e.name AS employee_name, e.role AS employee_role
        FROM asset_allocations aa
        JOIN employees e ON aa.employee_id = e.id
        WHERE aa.asset_id = ?
        ORDER BY aa.allocated_at DESC
    """, (asset_dict['id'],)).fetchall()
    asset_dict['allocation_history'] = _rows_to_dicts(allocations)

    maintenance = conn.execute("""
        SELECT * FROM maintenance_requests
        WHERE asset_id = ?
        ORDER BY raised_at DESC
    """, (asset_dict['id'],)).fetchall()
    asset_dict['maintenance_history'] = _rows_to_dicts(maintenance)

    conn.close()
    return asset_dict


# ───────────────────────────────────────────────────────────
# 3. who_has_asset
# ───────────────────────────────────────────────────────────
def who_has_asset(db_path, asset_tag):
    """Find who currently holds a specific asset."""
    conn = _get_conn(db_path)

    result = conn.execute("""
        SELECT a.asset_tag, a.name AS asset_name, a.status,
               e.name AS employee_name, e.role AS employee_role,
               d.name AS department_name, p.name AS plant_name,
               aa.allocated_at, aa.expected_return, aa.status AS alloc_status
        FROM asset_allocations aa
        JOIN assets a ON aa.asset_id = a.id
        JOIN employees e ON aa.employee_id = e.id
        JOIN departments d ON e.department_id = d.id
        JOIN plants p ON a.plant_id = p.id
        WHERE a.asset_tag = ? AND aa.actual_return IS NULL AND aa.status = 'Active'
        ORDER BY aa.allocated_at DESC
        LIMIT 1
    """, (asset_tag,)).fetchone()

    conn.close()
    if result:
        return dict(result)
    return {"message": f"Asset '{asset_tag}' is not currently allocated to anyone"}


# ───────────────────────────────────────────────────────────
# 4. get_employee_assets
# ───────────────────────────────────────────────────────────
def get_employee_assets(db_path, employee_name):
    """All assets currently held by an employee (fuzzy match on name)."""
    conn = _get_conn(db_path)

    rows = conn.execute("""
        SELECT e.name AS employee_name, e.role,
               a.asset_tag, a.name AS asset_name, a.category,
               aa.allocated_at, aa.expected_return, aa.status AS alloc_status,
               d.name AS department_name, p.name AS plant_name
        FROM asset_allocations aa
        JOIN employees e ON aa.employee_id = e.id
        JOIN assets a ON aa.asset_id = a.id
        JOIN departments d ON e.department_id = d.id
        JOIN plants p ON a.plant_id = p.id
        WHERE e.name LIKE ? AND aa.actual_return IS NULL AND aa.status = 'Active'
        ORDER BY aa.allocated_at DESC
    """, (f"%{employee_name}%",)).fetchall()

    conn.close()
    if rows:
        return _rows_to_dicts(rows)
    return {"message": f"No active allocations found for employee matching '{employee_name}'"}


# ───────────────────────────────────────────────────────────
# 5. count_assets
# ───────────────────────────────────────────────────────────
def count_assets(db_path, category=None, plant=None, status=None,
                 department=None, is_production_critical=None):
    """Count assets matching the given filters."""
    conn = _get_conn(db_path)
    sql = """
        SELECT COUNT(*) as count
        FROM assets a
        JOIN plants p ON a.plant_id = p.id
        JOIN departments d ON a.department_id = d.id
        WHERE 1=1
    """
    params = []
    if category:
        sql += " AND a.category = ?"
        params.append(category)
    if plant:
        sql += " AND (p.name LIKE ? OR p.city LIKE ?)"
        params.extend([f"%{plant}%", f"%{plant}%"])
    if status:
        sql += " AND a.status = ?"
        params.append(status)
    if department:
        sql += " AND d.name LIKE ?"
        params.append(f"%{department}%")
    if is_production_critical is not None:
        sql += " AND a.is_production_critical = ?"
        params.append(int(is_production_critical))

    result = conn.execute(sql, params).fetchone()
    conn.close()
    return {"count": result['count']}


# ───────────────────────────────────────────────────────────
# 6. list_maintenance_requests
# ───────────────────────────────────────────────────────────
def list_maintenance_requests(db_path, status=None, priority=None, category=None):
    """List maintenance requests with optional filters."""
    conn = _get_conn(db_path)
    sql = """
        SELECT mr.id, mr.issue_description, mr.priority, mr.status,
               mr.raised_at, mr.resolved_at,
               a.asset_tag, a.name AS asset_name, a.category,
               p.name AS plant_name
        FROM maintenance_requests mr
        JOIN assets a ON mr.asset_id = a.id
        JOIN plants p ON a.plant_id = p.id
        WHERE 1=1
    """
    params = []
    if status:
        sql += " AND mr.status = ?"
        params.append(status)
    if priority:
        sql += " AND mr.priority = ?"
        params.append(priority)
    if category:
        sql += " AND a.category = ?"
        params.append(category)
    sql += " ORDER BY mr.raised_at DESC LIMIT 50"

    rows = conn.execute(sql, params).fetchall()
    conn.close()
    return _rows_to_dicts(rows)


# ───────────────────────────────────────────────────────────
# 7. get_overdue_allocations
# ───────────────────────────────────────────────────────────
def get_overdue_allocations(db_path):
    """Allocations past expected_return with no actual_return."""
    conn = _get_conn(db_path)
    now = datetime.now().strftime('%Y-%m-%dT%H:%M:%S')

    rows = conn.execute("""
        SELECT aa.id, a.asset_tag, a.name AS asset_name, a.category,
               e.name AS employee_name, e.role AS employee_role,
               d.name AS department_name, p.name AS plant_name,
               aa.allocated_at, aa.expected_return,
               CAST(julianday(?) - julianday(aa.expected_return) AS INTEGER) AS days_overdue
        FROM asset_allocations aa
        JOIN assets a ON aa.asset_id = a.id
        JOIN employees e ON aa.employee_id = e.id
        JOIN departments d ON e.department_id = d.id
        JOIN plants p ON a.plant_id = p.id
        WHERE aa.actual_return IS NULL
          AND aa.expected_return < ?
          AND aa.status = 'Active'
        ORDER BY aa.expected_return ASC
    """, (now, now)).fetchall()

    conn.close()
    return _rows_to_dicts(rows)


# ───────────────────────────────────────────────────────────
# 8. get_upcoming_bookings
# ───────────────────────────────────────────────────────────
def get_upcoming_bookings(db_path, asset_tag=None, employee_name=None):
    """Get upcoming/active bookings with optional filters."""
    conn = _get_conn(db_path)
    sql = """
        SELECT b.id, b.start_time, b.end_time, b.status AS booking_status,
               a.asset_tag, a.name AS asset_name, a.category,
               e.name AS employee_name,
               p.name AS plant_name
        FROM bookings b
        JOIN assets a ON b.asset_id = a.id
        JOIN employees e ON b.employee_id = e.id
        JOIN plants p ON a.plant_id = p.id
        WHERE b.status IN ('Confirmed', 'Pending')
    """
    params = []
    if asset_tag:
        sql += " AND a.asset_tag = ?"
        params.append(asset_tag)
    if employee_name:
        sql += " AND e.name LIKE ?"
        params.append(f"%{employee_name}%")
    sql += " ORDER BY b.start_time ASC LIMIT 50"

    rows = conn.execute(sql, params).fetchall()
    conn.close()
    return _rows_to_dicts(rows)


# ───────────────────────────────────────────────────────────
# 9. get_department_summary
# ───────────────────────────────────────────────────────────
def get_department_summary(db_path, department_name):
    """Asset counts, employee counts, and maintenance stats for a department."""
    conn = _get_conn(db_path)

    dept = conn.execute(
        "SELECT d.*, p.name AS plant_name FROM departments d JOIN plants p ON d.plant_id = p.id WHERE d.name LIKE ?",
        (f"%{department_name}%",)
    ).fetchone()

    if not dept:
        conn.close()
        return {"error": f"Department matching '{department_name}' not found"}

    dept_dict = dict(dept)
    dept_id = dept_dict['id']

    # Employee count
    emp_count = conn.execute(
        "SELECT COUNT(*) as count FROM employees WHERE department_id = ?", (dept_id,)
    ).fetchone()['count']

    # Asset counts by status
    asset_stats = conn.execute("""
        SELECT status, COUNT(*) as count FROM assets
        WHERE department_id = ? GROUP BY status
    """, (dept_id,)).fetchall()

    # Open maintenance requests
    maint_stats = conn.execute("""
        SELECT mr.status, COUNT(*) as count
        FROM maintenance_requests mr
        JOIN assets a ON mr.asset_id = a.id
        WHERE a.department_id = ?
        GROUP BY mr.status
    """, (dept_id,)).fetchall()

    conn.close()
    return {
        "department": dept_dict,
        "employee_count": emp_count,
        "assets_by_status": _rows_to_dicts(asset_stats),
        "maintenance_by_status": _rows_to_dicts(maint_stats),
    }


# ───────────────────────────────────────────────────────────
# 10. run_custom_query
# ───────────────────────────────────────────────────────────
_FORBIDDEN_KEYWORDS = re.compile(
    r'\b(INSERT|UPDATE|DELETE|DROP|ALTER|CREATE|REPLACE|TRUNCATE|ATTACH|DETACH|PRAGMA|GRANT|REVOKE)\b',
    re.IGNORECASE,
)


def run_custom_query(db_path, sql):
    """Execute a read-only SQL query. Only SELECT statements are allowed."""
    stripped = sql.strip().rstrip(';').strip()
    if not stripped.upper().startswith('SELECT'):
        return {"error": "Only SELECT queries are allowed"}
    if _FORBIDDEN_KEYWORDS.search(stripped):
        return {"error": "Query contains forbidden keywords (write/DDL operations are not allowed)"}

    conn = _get_conn(db_path)
    try:
        if 'LIMIT' not in stripped.upper():
            stripped += " LIMIT 50"
        rows = conn.execute(stripped).fetchall()
        return _rows_to_dicts(rows)
    except Exception as e:
        return {"error": str(e)}
    finally:
        conn.close()


# ───────────────────────────────────────────────────────────
# TOOL_FUNCTIONS mapping
# ───────────────────────────────────────────────────────────
TOOL_FUNCTIONS = {
    'query_assets': query_assets,
    'get_asset_detail': get_asset_detail,
    'who_has_asset': who_has_asset,
    'get_employee_assets': get_employee_assets,
    'count_assets': count_assets,
    'list_maintenance_requests': list_maintenance_requests,
    'get_overdue_allocations': get_overdue_allocations,
    'get_upcoming_bookings': get_upcoming_bookings,
    'get_department_summary': get_department_summary,
    'run_custom_query': run_custom_query,
}


# ───────────────────────────────────────────────────────────
# TOOL_SCHEMAS (Ollama-compatible)
# ───────────────────────────────────────────────────────────
TOOL_SCHEMAS = [
    {
        "type": "function",
        "function": {
            "name": "query_assets",
            "description": "Search and list assets with optional filters. Returns asset details including tag, name, category, status, condition, location, plant, and department. Use this to find assets matching specific criteria.",
            "parameters": {
                "type": "object",
                "properties": {
                    "category": {
                        "type": "string",
                        "description": "Filter by asset category. Options: CNC Machine, Forklift, Conveyor Belt, Compressor, Generator, Laptop, Meeting Room, Vehicle, Welding Station, 3D Printer"
                    },
                    "plant": {
                        "type": "string",
                        "description": "Filter by plant name or city (partial match). E.g. 'Detroit', 'Austin', 'Chicago'"
                    },
                    "status": {
                        "type": "string",
                        "description": "Filter by asset status. Options: Available, Allocated, Under Maintenance, Reserved, Retired"
                    },
                    "department": {
                        "type": "string",
                        "description": "Filter by department name (partial match). E.g. 'Assembly', 'IT', 'Logistics'"
                    },
                    "is_production_critical": {
                        "type": "boolean",
                        "description": "Filter by whether the asset is production-critical (true/false)"
                    }
                },
                "required": []
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "get_asset_detail",
            "description": "Get complete details for a single asset by its asset tag, including full allocation history and maintenance history. Use this when you need in-depth information about a specific asset.",
            "parameters": {
                "type": "object",
                "properties": {
                    "asset_tag": {
                        "type": "string",
                        "description": "The unique asset tag identifier, e.g. 'CNC-DET-001', 'FLT-AUS-002'"
                    }
                },
                "required": ["asset_tag"]
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "who_has_asset",
            "description": "Find out who currently holds/is allocated a specific asset. Returns the employee name, role, department, and allocation details.",
            "parameters": {
                "type": "object",
                "properties": {
                    "asset_tag": {
                        "type": "string",
                        "description": "The unique asset tag identifier, e.g. 'CNC-DET-001'"
                    }
                },
                "required": ["asset_tag"]
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "get_employee_assets",
            "description": "Get all assets currently allocated to a specific employee. Uses fuzzy name matching so partial names work. Returns asset details and allocation info.",
            "parameters": {
                "type": "object",
                "properties": {
                    "employee_name": {
                        "type": "string",
                        "description": "Full or partial employee name to search for, e.g. 'Maria', 'Johnson', 'Carlos Rodriguez'"
                    }
                },
                "required": ["employee_name"]
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "count_assets",
            "description": "Get a count of assets matching the given filters. Use this when the user just wants to know how many assets match certain criteria rather than seeing all the details.",
            "parameters": {
                "type": "object",
                "properties": {
                    "category": {
                        "type": "string",
                        "description": "Filter by asset category. Options: CNC Machine, Forklift, Conveyor Belt, Compressor, Generator, Laptop, Meeting Room, Vehicle, Welding Station, 3D Printer"
                    },
                    "plant": {
                        "type": "string",
                        "description": "Filter by plant name or city (partial match)"
                    },
                    "status": {
                        "type": "string",
                        "description": "Filter by asset status. Options: Available, Allocated, Under Maintenance, Reserved, Retired"
                    },
                    "department": {
                        "type": "string",
                        "description": "Filter by department name (partial match)"
                    },
                    "is_production_critical": {
                        "type": "boolean",
                        "description": "Filter by whether the asset is production-critical"
                    }
                },
                "required": []
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "list_maintenance_requests",
            "description": "List maintenance requests with optional filters on status, priority, and asset category. Returns issue descriptions, priorities, statuses, dates, and associated asset info.",
            "parameters": {
                "type": "object",
                "properties": {
                    "status": {
                        "type": "string",
                        "description": "Filter by request status. Options: Open, In Progress, Resolved, Cancelled"
                    },
                    "priority": {
                        "type": "string",
                        "description": "Filter by priority level. Options: Low, Medium, High, Critical"
                    },
                    "category": {
                        "type": "string",
                        "description": "Filter by asset category to see maintenance for specific asset types"
                    }
                },
                "required": []
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "get_overdue_allocations",
            "description": "Find all asset allocations that are overdue - past their expected return date with no actual return recorded. Returns employee info, asset info, and how many days overdue.",
            "parameters": {
                "type": "object",
                "properties": {},
                "required": []
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "get_upcoming_bookings",
            "description": "Get upcoming and active bookings for bookable assets (meeting rooms, vehicles, forklifts, 3D printers). Can filter by asset tag or employee name.",
            "parameters": {
                "type": "object",
                "properties": {
                    "asset_tag": {
                        "type": "string",
                        "description": "Filter bookings for a specific asset tag"
                    },
                    "employee_name": {
                        "type": "string",
                        "description": "Filter bookings by employee name (partial match)"
                    }
                },
                "required": []
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "get_department_summary",
            "description": "Get a summary of a department including employee count, asset counts by status, and maintenance request statistics. Use for department-level overviews.",
            "parameters": {
                "type": "object",
                "properties": {
                    "department_name": {
                        "type": "string",
                        "description": "Department name or partial name, e.g. 'Assembly', 'IT', 'Logistics'"
                    }
                },
                "required": ["department_name"]
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "run_custom_query",
            "description": "Execute a custom read-only SQL SELECT query against the database. Use this for complex queries that cannot be answered by other tools. Only SELECT statements are allowed; write operations are blocked. Results limited to 50 rows. Tables: plants, departments, employees, assets, asset_allocations, maintenance_requests, bookings.",
            "parameters": {
                "type": "object",
                "properties": {
                    "sql": {
                        "type": "string",
                        "description": "A SQL SELECT query to execute. Must start with SELECT. Example: SELECT a.asset_tag, a.name FROM assets a WHERE a.category = 'Forklift'"
                    }
                },
                "required": ["sql"]
            }
        }
    },
]

File: server/modules/ai/tools.js
const prisma = require('../../shared/prismaClient');

/** Search/filter assets */
async function queryAssets({ category, status, location, search, limit = 20 }) {
  try {
    const where = {};
    if (category) where.category = { name: { contains: category } };
    if (status) where.status = status;
    if (location) where.location = { contains: location };
    if (search) where.OR = [{ name: { contains: search } }, { asset_tag: { contains: search } }];
    const assets = await prisma.asset.findMany({ where, take: parseInt(limit), include: { category: true }, orderBy: { id: 'asc' } });
    return assets.map(a => ({ id: a.id, asset_tag: a.asset_tag, name: a.name, category: a.category?.name, status: a.status, condition: a.condition, location: a.location, is_bookable: a.is_bookable }));
  } catch (e) { return { error: e.message }; }
}

/** Full detail on one asset */
async function getAssetDetail({ asset_tag }) {
  try {
    const asset = await prisma.asset.findUnique({ where: { asset_tag }, include: { category: true, allocations: { include: { holder_user: true, allocator: true }, orderBy: { allocated_at: 'desc' }, take: 5 }, maintenance_requests: { orderBy: { created_at: 'desc' }, take: 5 } } });
    if (!asset) return { error: 'Asset not found' };
    return { id: asset.id, asset_tag: asset.asset_tag, name: asset.name, category: asset.category?.name, status: asset.status, condition: asset.condition, location: asset.location, acquisition_cost: asset.acquisition_cost, acquisition_date: asset.acquisition_date, current_allocation: asset.allocations.find(a => a.status === 'Active'), allocation_count: asset.allocations.length, maintenance_count: asset.maintenance_requests.length, recent_maintenance: asset.maintenance_requests.slice(0, 3).map(m => ({ issue: m.issue_description, priority: m.priority, status: m.status })) };
  } catch (e) { return { error: e.message }; }
}

/** Who currently holds an asset */
async function whoHasAsset({ asset_tag }) {
  try {
    const asset = await prisma.asset.findUnique({ where: { asset_tag }, include: { allocations: { where: { status: 'Active' }, include: { holder_user: true, holder_department: true } } } });
    if (!asset) return { error: 'Asset not found' };
    const alloc = asset.allocations[0];
    if (!alloc) return { asset_tag, status: asset.status, holder: 'Nobody — asset is not currently allocated' };
    return { asset_tag, holder: alloc.holder_user?.name || alloc.holder_department?.name || 'Unknown', allocated_since: alloc.allocated_at, expected_return: alloc.expected_return_date };
  } catch (e) { return { error: e.message }; }
}

/** All assets held by an employee */
async function getEmployeeAssets({ employee_name }) {
  try {
    const users = await prisma.user.findMany({ where: { name: { contains: employee_name } } });
    if (users.length === 0) return { error: `No employee found matching "${employee_name}"` };
    const allocations = await prisma.allocation.findMany({ where: { holder_user_id: { in: users.map(u => u.id) }, status: 'Active' }, include: { asset: { include: { category: true } }, holder_user: true } });
    return allocations.map(a => ({ employee: a.holder_user?.name, asset: a.asset?.name, asset_tag: a.asset?.asset_tag, category: a.asset?.category?.name, since: a.allocated_at }));
  } catch (e) { return { error: e.message }; }
}

/** Count matching assets */
async function countAssets({ category, status, location }) {
  try {
    const where = {};
    if (category) where.category = { name: { contains: category } };
    if (status) where.status = status;
    if (location) where.location = { contains: location };
    const count = await prisma.asset.count({ where });
    return { count, filters: { category, status, location } };
  } catch (e) { return { error: e.message }; }
}

/** List maintenance requests */
async function listMaintenanceRequests({ status, priority }) {
  try {
    const where = {};
    if (status) where.status = status;
    if (priority) where.priority = priority;
    const requests = await prisma.maintenanceRequest.findMany({ where, take: 20, include: { asset: true, raiser: true }, orderBy: { created_at: 'desc' } });
    return requests.map(r => ({ id: r.id, asset: r.asset?.name, asset_tag: r.asset?.asset_tag, issue: r.issue_description, priority: r.priority, status: r.status, raised_by: r.raiser?.name, date: r.created_at }));
  } catch (e) { return { error: e.message }; }
}

/** Get overdue allocations */
async function getOverdueAllocations() {
  try {
    const overdue = await prisma.allocation.findMany({ where: { status: 'Active', expected_return_date: { lt: new Date() } }, include: { asset: true, holder_user: true } });
    return overdue.map(a => ({ asset: a.asset?.name, asset_tag: a.asset?.asset_tag, holder: a.holder_user?.name, expected_return: a.expected_return_date, days_overdue: Math.floor((Date.now() - new Date(a.expected_return_date)) / 86400000) }));
  } catch (e) { return { error: e.message }; }
}

/** Get upcoming bookings */
async function getUpcomingBookings({ asset_tag, employee_name }) {
  try {
    const where = { start_time: { gte: new Date() } };
    if (asset_tag) where.asset = { asset_tag };
    const bookings = await prisma.booking.findMany({ where, take: 20, include: { asset: true, user: true }, orderBy: { start_time: 'asc' } });
    let result = bookings.map(b => ({ asset: b.asset?.name, asset_tag: b.asset?.asset_tag, booked_by: b.user?.name, start: b.start_time, end: b.end_time, purpose: b.purpose, status: b.status }));
    if (employee_name) result = result.filter(b => b.booked_by?.toLowerCase().includes(employee_name.toLowerCase()));
    return result;
  } catch (e) { return { error: e.message }; }
}

/** Department overview */
async function getDepartmentSummary({ department_name }) {
  try {
    const dept = await prisma.department.findFirst({ where: { name: { contains: department_name } }, include: { users: true } });
    if (!dept) return { error: `Department "${department_name}" not found` };
    const userIds = dept.users.map(u => u.id);
    const [activeAllocs, overdueAllocs, maintReqs] = await Promise.all([
      prisma.allocation.count({ where: { holder_user_id: { in: userIds }, status: 'Active' } }),
      prisma.allocation.count({ where: { holder_user_id: { in: userIds }, status: 'Active', expected_return_date: { lt: new Date() } } }),
      prisma.maintenanceRequest.count({ where: { raised_by: { in: userIds }, status: { in: ['Pending', 'Approved', 'InProgress'] } } }),
    ]);
    return { department: dept.name, employee_count: dept.users.length, employees: dept.users.map(u => u.name), active_allocations: activeAllocs, overdue_allocations: overdueAllocs, open_maintenance: maintReqs };
  } catch (e) { return { error: e.message }; }
}

/** Dashboard stats */
async function getDashboardStats() {
  try {
    const [total, available, allocated, maintenance, overdue] = await Promise.all([
      prisma.asset.count(),
      prisma.asset.count({ where: { status: 'Available' } }),
      prisma.asset.count({ where: { status: 'Allocated' } }),
      prisma.asset.count({ where: { status: 'In Maintenance' } }),
      prisma.allocation.count({ where: { status: 'Active', expected_return_date: { lt: new Date() } } }),
    ]);
    return { total_assets: total, available, allocated, in_maintenance: maintenance, overdue_returns: overdue };
  } catch (e) { return { error: e.message }; }
}

/** Search across assets, users, departments */
async function searchEverything({ query }) {
  try {
    const [assets, users, depts] = await Promise.all([
      prisma.asset.findMany({ where: { OR: [{ name: { contains: query } }, { asset_tag: { contains: query } }] }, take: 5, select: { id: true, name: true, asset_tag: true, status: true } }),
      prisma.user.findMany({ where: { name: { contains: query } }, take: 5, select: { id: true, name: true, email: true, role: true } }),
      prisma.department.findMany({ where: { name: { contains: query } }, take: 5, select: { id: true, name: true } }),
    ]);
    return { assets, users, departments: depts };
  } catch (e) { return { error: e.message }; }
}

const TOOL_SCHEMAS = [
  { type: 'function', function: { name: 'query_assets', description: 'Search and filter assets by category, status, location, or search term', parameters: { type: 'object', properties: { category: { type: 'string', description: 'Asset category name' }, status: { type: 'string', description: 'Asset status', enum: ['Available', 'Allocated', 'In Maintenance', 'Retired', 'Lost'] }, location: { type: 'string' }, search: { type: 'string' }, limit: { type: 'integer', description: 'Max results (default 20)' } } } } },
  { type: 'function', function: { name: 'get_asset_detail', description: 'Get full details on a specific asset including allocation and maintenance history', parameters: { type: 'object', properties: { asset_tag: { type: 'string', description: 'The asset tag (e.g. AF-0001)' } }, required: ['asset_tag'] } } },
  { type: 'function', function: { name: 'who_has_asset', description: 'Find out who currently holds a specific asset', parameters: { type: 'object', properties: { asset_tag: { type: 'string' } }, required: ['asset_tag'] } } },
  { type: 'function', function: { name: 'get_employee_assets', description: 'Get all assets currently held by an employee (fuzzy name match)', parameters: { type: 'object', properties: { employee_name: { type: 'string' } }, required: ['employee_name'] } } },
  { type: 'function', function: { name: 'count_assets', description: 'Count assets matching filters', parameters: { type: 'object', properties: { category: { type: 'string' }, status: { type: 'string' }, location: { type: 'string' } } } } },
  { type: 'function', function: { name: 'list_maintenance_requests', description: 'List maintenance requests, optionally filtered by status and priority', parameters: { type: 'object', properties: { status: { type: 'string', enum: ['Pending', 'Approved', 'InProgress', 'Resolved', 'Rejected'] }, priority: { type: 'string', enum: ['Low', 'Medium', 'High', 'Critical'] } } } } },
  { type: 'function', function: { name: 'get_overdue_allocations', description: 'Get all assets that are past their expected return date', parameters: { type: 'object', properties: {} } } },
  { type: 'function', function: { name: 'get_upcoming_bookings', description: 'Get upcoming asset bookings', parameters: { type: 'object', properties: { asset_tag: { type: 'string' }, employee_name: { type: 'string' } } } } },
  { type: 'function', function: { name: 'get_department_summary', description: 'Get an overview of a department including employees, assets, and maintenance', parameters: { type: 'object', properties: { department_name: { type: 'string' } }, required: ['department_name'] } } },
  { type: 'function', function: { name: 'get_dashboard_stats', description: 'Get overall dashboard statistics', parameters: { type: 'object', properties: {} } } },
  { type: 'function', function: { name: 'search_everything', description: 'Search across assets, users, and departments', parameters: { type: 'object', properties: { query: { type: 'string' } }, required: ['query'] } } },
];

const TOOL_FUNCTIONS = { query_assets: queryAssets, get_asset_detail: getAssetDetail, who_has_asset: whoHasAsset, get_employee_assets: getEmployeeAssets, count_assets: countAssets, list_maintenance_requests: listMaintenanceRequests, get_overdue_allocations: getOverdueAllocations, get_upcoming_bookings: getUpcomingBookings, get_department_summary: getDepartmentSummary, get_dashboard_stats: getDashboardStats, search_everything: searchEverything };

module.exports = { TOOL_SCHEMAS, TOOL_FUNCTIONS };
